### ISSUE TITLE
all(isintance(....)) filters types in python but pylance doesn't recognize it

### ISSUE THREAD

**Issue by RS-Credentive** [microsoft/pyright#9683]
## Environment data

-   Pylance version: 2024.12.1
-   OS and version: dev container (python:1-3.12-bullseye)
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.8

## Code Snippet

```python
from typing import TypeAlias

type_t: TypeAlias = dict[str, bool | list[str]]

d1: type_t = {"1": False, "2": True, "3": False}
d2: type_t = {"2": ["a", "b", "c"]}


if all([isinstance(v, list) for v in d1.values()]):
    for v in d1.values():
        print(v)
else:
    print("d1 contains non list values")

if all([isinstance(v, list) for v in d2.values()]):
    for l in d2.values():
        for item in l:  # "bool" is not iterable
                        # "__iter__" method not defined Pylance
            print(item)

# OUTPUT:
# d1 contains non list values
# a
# b
# c
```

## Expected behavior

since all(isinstance()) succesfully filters types to ensure they are all lists, pylance should not complain about the looping construct in the second if 

## Actual behavior

pylance complains about the line "for item in l" that a bool is not iterable

## Logs

```
2025-01-10 11:11:50.050 [info] [Info  - 11:11:50 AM] (41693) Server root directory: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist
2025-01-10 11:11:50.050 [info] [Info  - 11:11:50 AM] (41693) Pylance language server 2024.12.1 (pyright version 1.1.389, commit ce1325fc) starting
2025-01-10 11:11:50.053 [info] [Info  - 11:11:50 AM] (41693) Starting service instance "pylance_test"
2025-01-10 11:11:50.077 [info] (41693) No configuration file found.
2025-01-10 11:11:50.077 [info] (41693) No pyproject.toml file found.
2025-01-10 11:11:50.077 [info] [Info  - 11:11:50 AM] (41693) Setting environmentName for service "pylance_test": "3.12.8 (global)"
2025-01-10 11:11:50.078 [info] [Info  - 11:11:50 AM] (41693) Setting pythonPath for service "pylance_test": "/usr/local/bin/python"
2025-01-10 11:11:50.078 [info] [Info  - 11:11:50 AM] (41693) No include entries specified; assuming /home/vscode/pylance_test
2025-01-10 11:11:50.078 [info] [Info  - 11:11:50 AM] (41693) Auto-excluding **/node_modules
2025-01-10 11:11:50.078 [info] [Info  - 11:11:50 AM] (41693) Auto-excluding **/__pycache__
2025-01-10 11:11:50.078 [info] [Info  - 11:11:50 AM] (41693) Auto-excluding **/.*
2025-01-10 11:11:50.090 [info] [Info  - 11:11:50 AM] (41693) Assuming Python version 3.12.8.final.0
2025-01-10 11:11:50.090 [info] (41693) Assuming Python platform Linux
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693) Execution environment: 3.12.8 (global)
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)   Extra paths:
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     (none)
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)   Python version: 3.12.8.final.0
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)   Python platform: Linux
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)   Search paths:
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /home/vscode/pylance_test
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /home/vscode/pylance_test/typings
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stubs/...
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/bundled/stubs
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /usr/local/lib/python3.12
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /usr/local/lib/python3.12/lib-dynload
2025-01-10 11:11:50.116 [info] [Info  - 11:11:50 AM] (41693)     /usr/local/lib/python3.12/site-packages
2025-01-10 11:11:50.117 [info] [Info  - 11:11:50 AM] (41693) Adding fs watcher for library directories:
 file:///usr/local/lib/python3.12
2025-01-10 11:11:50.117 [info] [Info  - 11:11:50 AM] (41693) Adding fs watcher for directories:
 file:///home/vscode/pylance_test
2025-01-10 11:11:50.117 [info] (41693) Searching for source files
2025-01-10 11:11:50.119 [info] [Info  - 11:11:50 AM] (41693) Found 1 source file
2025-01-10 11:11:50.119 [info] (41693) Adding fs watcher for files:
 file:///home/vscode/pylance_test/requirements.txt
file:///home/vscode/pylance_test/pyproject.toml
2025-01-10 11:11:50.124 [info] (41693) pytest configurations: {"message":"request cancelled","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2025-01-10 11:11:50.139 [info] (41693) Attempting to resolve using local imports: __builtins__
2025-01-10 11:11:50.142 [info] (41693) [FG] parsing: file:///home/vscode/pylance_test/typing_test.py (18ms)
2025-01-10 11:11:50.188 [info] (41693) [FG] parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 1ms] (45ms)
2025-01-10 11:11:50.208 [info] (41693) [FG] binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/builtins.pyi (20ms)
2025-01-10 11:11:50.209 [info] (41693) [FG] binding: file:///home/vscode/pylance_test/typing_test.py (1ms)
2025-01-10 11:11:50.215 [info] (41693) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2025-01-10 11:11:50.289 [info] [Info  - 11:11:50 AM] (41693) Background analysis(1) root directory: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist
2025-01-10 11:11:50.291 [info] [Info  - 11:11:50 AM] (41693) Background analysis(1) started
2025-01-10 11:11:50.321 [info] [Info  - 11:11:50 AM] (41693) Indexer background runner(2) root directory: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist (index)
2025-01-10 11:11:50.321 [info] [Info  - 11:11:50 AM] (41693) Indexing(2) started
2025-01-10 11:11:50.323 [info] (41693) pytest configurations: {"message":"request cancelled","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2025-01-10 11:11:50.329 [info] (41693) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
2025-01-10 11:11:50.358 [info] (41693) [IDX(2)] scan packages file:///home/vscode/pylance_test ...
2025-01-10 11:11:50.358 [info] (41693) [IDX(2)]   read stdlib indices (10ms)
2025-01-10 11:11:50.465 [info] (41693) [IDX(2)] scan packages file:///home/vscode/pylance_test (117ms)
2025-01-10 11:11:50.465 [info] [Info  - 11:11:50 AM] (41693) scanned(2) 17 files over 1 exec env
2025-01-10 11:11:50.558 [info] (41693) [IDX(2)] index packages file:///home/vscode/pylance_test ...
2025-01-10 11:11:50.559 [info] (41693) [IDX(2)]   index execution environment file:///home/vscode/pylance_test [found 926 in 17 files] (2ms)
2025-01-10 11:11:50.559 [info] (41693) [IDX(2)] index packages file:///home/vscode/pylance_test [found 926 in 1 exec envs] (6ms)
2025-01-10 11:11:50.559 [info] [Info  - 11:11:50 AM] (41693) indexed(2) 17 files over 1 exec env
2025-01-10 11:11:50.569 [info] [Info  - 11:11:50 AM] (41693) Indexing finished(2).
2025-01-10 11:11:50.584 [info] (41693) Attempting to resolve using local imports: __builtins__
2025-01-10 11:11:50.585 [info] (41693) [BG(1)] SemanticTokens full at file:///home/vscode/pylance_test/typing_test.py ...
2025-01-10 11:11:50.585 [info] (41693) [BG(1)]   parsing: file:///home/vscode/pylance_test/typing_test.py (21ms)
2025-01-10 11:11:50.626 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 1ms] (40ms)
2025-01-10 11:11:50.647 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/builtins.pyi (20ms)
2025-01-10 11:11:50.648 [info] (41693) [BG(1)]   binding: file:///home/vscode/pylance_test/typing_test.py (1ms)
2025-01-10 11:11:50.660 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/typing.pyi [fs read 1ms] (10ms)
2025-01-10 11:11:50.668 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/typing.pyi (8ms)
2025-01-10 11:11:50.684 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/types.pyi [fs read 0ms] (6ms)
2025-01-10 11:11:50.686 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/types.pyi (3ms)
2025-01-10 11:11:50.693 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 0ms] (3ms)
2025-01-10 11:11:50.694 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (1ms)
2025-01-10 11:11:50.698 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/abc.pyi [fs read 0ms] (1ms)
2025-01-10 11:11:50.698 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/abc.pyi (0ms)
2025-01-10 11:11:50.702 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/collections/abc.pyi [fs read 0ms] (0ms)
2025-01-10 11:11:50.703 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/collections/abc.pyi ...
2025-01-10 11:11:50.703 [info] (41693) [BG(1)]     parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi [fs read 0ms] (0ms)
2025-01-10 11:11:50.703 [info] (41693) [BG(1)]     binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi (0ms)
2025-01-10 11:11:50.703 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/collections/abc.pyi (1ms)
2025-01-10 11:11:50.706 [info] (41693) [BG(1)]   parsing: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 0ms] (2ms)
2025-01-10 11:11:50.708 [info] (41693) [BG(1)]   binding: file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi (2ms)
2025-01-10 11:11:50.728 [info] (41693) [BG(1)] SemanticTokens full at file:///home/vscode/pylance_test/typing_test.py (165ms)
2025-01-10 11:11:50.729 [info] (41693) [BG(1)] SemanticTokens range 0:0 - 24:3 at file:///home/vscode/pylance_test/typing_test.py (1ms)
2025-01-10 11:11:50.753 [info] (41693) [BG(1)] indexing: file:///home/vscode/pylance_test/typing_test.py [found 6] (0ms)
2025-01-10 11:11:50.753 [info] (41693) Workspace indexing done: file:///home/vscode/pylance_test/typing_test.py
2025-01-10 11:11:50.786 [info] (41693) [BG(1)] analyzing: file:///home/vscode/pylance_test/typing_test.py ...
2025-01-10 11:11:50.786 [info] (41693) [BG(1)]   checking: file:///home/vscode/pylance_test/typing_test.py ...
2025-01-10 11:11:50.786 [info] (41693) [BG(1)]     parsing: file:///usr/local/lib/python3.12/typing.py [fs read 0ms] (30ms)
2025-01-10 11:11:50.786 [info] [Info  - 11:11:50 AM] (41693) Could not import '_typing' in file '/usr/local/lib/python3.12/typing.py'
2025-01-10 11:11:50.786 [info] [Info  - 11:11:50 AM] (41693)   Looking in stubPath 'file:///home/vscode/pylance_test/typings'
2025-01-10 11:11:50.786 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///home/vscode/pylance_test/typings'
2025-01-10 11:11:50.786 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///home/vscode/pylance_test/typings'
2025-01-10 11:11:50.786 [info] [Info  - 11:11:50 AM] (41693)   Looking in root directory of execution environment 'file:///home/vscode/pylance_test'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///home/vscode/pylance_test'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///home/vscode/pylance_test'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking for typeshed stdlib path
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking for typeshed stdlib path
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/typeshed-fallback/stdlib'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Typeshed path not found
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Finding python search paths
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Executing interpreter: '/usr/local/bin/python'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Skipping '/usr/local/lib/python312.zip' because it is not a valid directory
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Received 3 paths from interpreter
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)     file:///usr/local/lib/python3.12
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)     file:///usr/local/lib/python3.12/lib-dynload
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)     file:///usr/local/lib/python3.12/site-packages
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking in python search path 'file:///usr/local/lib/python3.12'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///usr/local/lib/python3.12'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///usr/local/lib/python3.12'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking in python search path 'file:///usr/local/lib/python3.12/lib-dynload'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///usr/local/lib/python3.12/lib-dynload'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///usr/local/lib/python3.12/lib-dynload'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking in python search path 'file:///usr/local/lib/python3.12/site-packages'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///usr/local/lib/python3.12/site-packages'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///usr/local/lib/python3.12/site-packages'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking in bundled stubs path 'file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/bundled/stubs'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve stub package using root path 'file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/bundled/stubs'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Attempting to resolve using root path 'file:///home/vscode/.vscode-server/extensions/ms-python.vscode-pylance-2024.12.1/dist/bundled/stubs'
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking for typeshed third-party path
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Looking for typeshed stubs path
2025-01-10 11:11:50.787 [info] [Info  - 11:11:50 AM] (41693)   Typeshed path not found
2025-01-10 11:11:50.797 [info] (41693) [BG(1)]     binding: file:///usr/local/lib/python3.12/typing.py (11ms)
2025-01-10 11:11:50.802 [info] (41693) [BG(1)]   checking: file:///home/vscode/pylance_test/typing_test.py (49ms)
2025-01-10 11:11:50.802 [info] (41693) [BG(1)] analyzing: file:///home/vscode/pylance_test/typing_test.py (49ms)
2025-01-10 11:11:50.818 [info] (41693) [BG(1)] SemanticTokens delta previousResultId:1736507510648 at file:///home/vscode/pylance_test/typing_test.py (1ms)
2025-01-10 11:11:52.575 [info] [Info  - 11:11:52 AM] (41693) SourceFile: Received fs event 'change' for path '/home/vscode/pylance_test/typing_test.py'
2025-01-10 11:11:52.584 [info] (41693) [BG(1)] indexing: file:///home/vscode/pylance_test/typing_test.py [found 6] (1ms)
2025-01-10 11:11:52.584 [info] (41693) Workspace indexing done: file:///home/vscode/pylance_test/typing_test.py
2025-01-10 11:11:52.585 [info] [Info  - 11:11:52 AM] (41693) Heap stats: total_memory_size=7834MB, total_free_size=6227MB, total_heap_size=84MB, used_heap_size=54MB, cross_worker_used_heap_size=54MB, total_physical_size=84MB, total_available_size=2041MB, heap_size_limit=2096MB
```

**Comment by heejaechang** [comment]
transferred typing issue to pyright.

**Comment by erictraut** [comment]
This is not a supported type guard form. For a full list of supported type guard forms supported by pyright, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

Also see [this issue](https://github.com/microsoft/pyright/issues/6595). 

Type guard forms are very specific. To support this form, pyright would need to add support for `all([isinstance(x, T) for x in y.values()])`. It's unlikely we'd ever add such support because this form is too complex and highly specific to your code. 

For more complex type guards, you can write your own "user-defined type guard function" using the `TypeGuard` or `TypeIs` capabilities of the Python type system.

**Comment by RS-Credentive** [comment]
Thanks for the pointer to user-defined TypeGuard/TypeIs. I'd heard of them but hadn't dug into them. They are nifty!
